### PR TITLE
[SPARK-8126] [BUILD] Use custom temp directory during build.

### PR DESCRIPTION
**Pull request by vanzin**
Even with all the efforts to cleanup the temp directories created by
unit tests, Spark leaves a lot of garbage in /tmp after a test run.
This change overrides java.io.tmpdir to place those files under the
build directory instead.

After an sbt full unit test run, I was left with > 400 MB of temp
files. Since they're now under the build dir, it's much easier to
clean them up.

Also make a slight change to a unit test to make it not pollute the
source directory with test data.

### CODE REVIEW

**Comment by vanzin** [comment]
This reverts the revert and fixes the issue the original change caused.

**Comment by JoshRosen** [comment]
Can you explain what the root problem was?  Why did this depend on test ordering?

**Comment by vanzin** [comment]
The root problem was that the tmp directory was not created (I added code to both builds to explicitly do it). I guess depending on which tests run first the directory might be there (some tests use guava's `Files.createTempDirectory`, maybe that one creates the dir if it isn't there).

**Comment by vanzin** [comment]
Jenkins, retest this please.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -51,6 +51,11 @@ object BuildCommons {
   // Root project.
   val spark = ProjectRef(buildLocation, "spark")
   val sparkHome = buildLocation
+
+  val testTempDir = s"$sparkHome/target/tmp"
+  if (!new File(testTempDir).isDirectory()) {
+    require(new File(testTempDir).mkdirs())
+  }
```
Comment (yhuai) [inline project/SparkBuild.scala]: cc @marmbrus

**Comment by srowen** [comment]
Looks good to me. Do we need any additional testing before trying to merge it again? I could merge to master and then later backport if it seems OK for a few days.

**Comment by andrewor14** [comment]
I've triggered a few more test runs for this patch. Hopefully this time it's not flaky.

**Comment by srowen** [comment]
Jenkins, retest this please

**Comment by srowen** [comment]
I'm going to merge this into master and see how it goes, then to 1.4 and see what happens, then 1.3.

**Comment by srowen** [comment]
`master` looks OK and has all day. Moving on to 1.4.

**Comment by srowen** [comment]
All of the 1.4 builds have succeeded since this patch, some a few times. The exception is: https://amplab.cs.berkeley.edu/jenkins/job/Spark-1.4-Maven-with-YARN/  This succeeded after, then failed, and the failure in the Kafka suite looks unrelated since it doesn't involve a temp file. I'm declaring victory and moving to 1.3.

File: project/SparkBuild.scala (hunk 1 above)
Comment (marmbrus) [inline project/SparkBuild.scala]: The problem with doing it this way is that `buildLocation` is just using the current directory as the build root, which is not always a valid assumption.  We should instead remove that `val` and use `projectRoot.value` anyplace that we need the root spark directory.

Comment (vanzin) [inline project/SparkBuild.scala]: That should affect also the `val spark =...` line above, right? So my particular change shouldn't have made anything worse than before.

Comment (marmbrus) [inline project/SparkBuild.scala]: Yeah, sorry I would have been more clear.  This was a general comment about how we are doing things in the SBT build and not something that should have blocked merging this PR.

**Pull request closed without merging**